ssntp/certs/certs.go

```diff
@@ -0,0 +1,240 @@
+//
+// Copyright (c) 2016 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Initial implementation based on
+//    golang/src/pkg/crypto/tls/generate_cert.go
+//
+// which is:
+//
+//    Copyright 2009 The Go Authors. All rights reserved.
+//    Use of this source code is governed by a BSD-style
+//    license that can be found in the golang LICENSE file.
+//
+
+package certs
+
+import (
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/asn1"
+	"encoding/pem"
+	"fmt"
+	"io"
+	"math/big"
+	"net"
+	"os"
+	"time"
+
+	"github.com/01org/ciao/ssntp"
+)
+
+func publicKey(priv interface{}) interface{} {
+	switch k := priv.(type) {
+	case *rsa.PrivateKey:
+		return &k.PublicKey
+	case *ecdsa.PrivateKey:
+		return &k.PublicKey
+	default:
+		return nil
+	}
+}
+
+func pemBlockForKey(priv interface{}) *pem.Block {
+	switch k := priv.(type) {
+	case *rsa.PrivateKey:
+		return &pem.Block{Type: "RSA PRIVATE KEY", Bytes: x509.MarshalPKCS1PrivateKey(k)}
+	case *ecdsa.PrivateKey:
+		b, err := x509.MarshalECPrivateKey(k)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Unable to marshal ECDSA private key: %v", err)
+			os.Exit(2)
+		}
+		return &pem.Block{Type: "EC PRIVATE KEY", Bytes: b}
+	default:
+		return nil
+	}
+}
+
+func keyFromPemBlock(block *pem.Block) (serverPrivKey interface{}, err error) {
+	if block.Type == "EC PRIVATE KEY" {
+		serverPrivKey, err = x509.ParseECPrivateKey(block.Bytes)
+	} else {
+		serverPrivKey, err = x509.ParsePKCS1PrivateKey(block.Bytes)
+	}
+	return
+}
+
+func addOIDs(role ssntp.Role, oids []asn1.ObjectIdentifier) []asn1.ObjectIdentifier {
+	if role.IsAgent() {
+		oids = append(oids, ssntp.RoleAgentOID)
+	}
+
+	if role.IsScheduler() {
+		oids = append(oids, ssntp.RoleSchedulerOID)
+	}
+
+	if role.IsController() {
+		oids = append(oids, ssntp.RoleControllerOID)
+	}
+
+	if role.IsNetAgent() {
+		oids = append(oids, ssntp.RoleNetAgentOID)
+	}
+
+	if role.IsServer() {
+		oids = append(oids, ssntp.RoleServerOID)
+	}
+
+	if role.IsCNCIAgent() {
+		oids = append(oids, ssntp.RoleCNCIAgentOID)
+	}
+
+	return oids
+}
+
+func generatePrivateKey(ell bool) (interface{}, error) {
+	var priv interface{}
+	var err error
+
+	if ell == false {
+		priv, err = rsa.GenerateKey(rand.Reader, 2048)
+	} else {
+		priv, err = ecdsa.GenerateKey(elliptic.P521(), rand.Reader)
+	}
+
+	if err != nil {
```

Comment:
Maybe we could make this function simpler.  If GenerateKey always sets priv to nil when there is an error, which it really should, we could just write.

```
func generatePrivateKey(ell bool) (interface{}, error) {
      if ell == false {
		return rsa.GenerateKey(rand.Reader, 2048)
	} else {
		return ecdsa.GenerateKey(elliptic.P521(), rand.Reader)
	}
}
```